flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/aftermatch/AfterMatchSkipStrategy.java

```diff
@@ -0,0 +1,155 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.cep.nfa.aftermatch;
+
+import org.apache.flink.cep.nfa.ComputationState;
+import org.apache.flink.cep.nfa.sharedbuffer.EventId;
+import org.apache.flink.cep.nfa.sharedbuffer.SharedBuffer;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+
+/**
+ * Indicate the skip strategy after a match process.
+ */
+public abstract class AfterMatchSkipStrategy implements Serializable {
+
+	private static final long serialVersionUID = -4048930333619068531L;
+
+	/**
+	 * Discards every partial match that contains event of the match preceding the first of *PatternName*.
+	 *
+	 * @param patternName the pattern name to skip to
+	 * @return the created AfterMatchSkipStrategy
+	 */
+	public static AfterMatchSkipStrategy skipToFirst(String patternName) {
+		return new SkipToFirstStrategy(patternName);
+	}
+
+	/**
+	 * Discards every partial match that contains event of the match preceding the last of *PatternName*.
+	 *
+	 * @param patternName the pattern name to skip to
+	 * @return the created AfterMatchSkipStrategy
+	 */
+	public static AfterMatchSkipStrategy skipToLast(String patternName) {
+		return new SkipToLastStrategy(patternName);
+	}
+
+	/**
+	 * Discards every partial match that contains event of the match.
+	 *
+	 * @return the created AfterMatchSkipStrategy
+	 */
+	public static AfterMatchSkipStrategy skipPastLastEvent() {
+		return SkipPastLastStrategy.INSTANCE;
+	}
+
+	/**
+	 * Every possible match will be emitted.
+	 *
+	 * @return the created AfterMatchSkipStrategy
+	 */
+	public static AfterMatchSkipStrategy noSkip() {
+		return NoSkipStrategy.INSTANCE;
+	}
+
+	/**
+	 * Tells if the strategy may skip some matches.
+	 *
+	 * @return false if the strategy is NO_SKIP strategy
+	 */
+	public abstract boolean isSkipStrategy();
+
+	/**
+	 * Prunes matches/partial matches based on the chosen strategy.
+	 *
+	 * @param partialMatches current partial matches
+	 * @param matchedResult  already completed matches
+	 * @param sharedBuffer   corresponding shared buffer
+	 * @throws Exception thrown if could not access the state
+	 */
+	public void prune(
+			Collection<ComputationState> partialMatches,
```

Comment:
The name `partialMatches` is misleading because we use it also with the `completedMatches`.